File: julia/pr037.jl
# Problem 37
#
# 14 February 2003
#
#
# The number 3797 has an interesting property. Being prime itself, it is
# possible to continuously remove digits from left to right, and remain
# prime at each stage: 3797, 797, 97, and 7. Similarly we can work from
# right to left: 3797, 379, 37, and 3.
#
# Find the sum of the only eleven primes that are both truncatable from left
# to right and right to left.
#
# NOTE: 2, 3, 5, and 7 are not considered to be truncatable primes.
#
# 748317

# Given a list of numbers, return a list of the numbers that are still
# prime when a single digit as appended to the right.
function add_primes{T}(numbers::Vector{T})
   result = T[]
   for n in numbers
      for extra in [1, 3, 7, 9]
         tmp = n * 10 + extra
         if isprime(tmp)
            push!(result, tmp)
         end
      end
   end
   result
end

# Generate a list of all right-truncatable primes.
function right_truncatable_primes()
   result = Int64[]
   work = [2, 3, 5, 7]

   while !isempty(work)
      append!(result, work)
      work = add_primes(work)
   end
   result
end

function numreverse(n)
   result = 0
   while n > 0
      result = result * 10 + mod(n, 10)
      n = div(n, 10)
   end
   result
end

left_truncate(n) = numreverse(div(numreverse(n), 10))

function is_left_truncatable(number)
   while number > 0
      if !isprime(number)
         return false
      end
      number = left_truncate(number)
   end
   true
end

function solve()
   rights = right_truncatable_primes()
   filter!(is_left_truncatable, rights)
   filter!(x -> x > 9, rights)
   sum(rights)
end

println(solve())

File: ocaml/pr037.ml
(*
 * Problem 37
 *
 * 14 February 2003
 *
 *
 * The number 3797 has an interesting property. Being prime itself, it
 * is possible to continuously remove digits from left to right, and
 * remain prime at each stage: 3797, 797, 97, and 7. Similarly we can
 * work from right to left: 3797, 379, 37, and 3.
 *
 * Find the sum of the only eleven primes that are both truncatable
 * from left to right and right to left.
 *
 * NOTE: 2, 3, 5, and 7 are not considered to be truncatable primes.
 *)

open! Batteries
open Printf

(*
module P = Sieve.Int64Factory

(* This prime function takes way too long.  Need to implement MR or
   something like that.  (This program is taking about 21 minutes) *)
let is_prime x = P.is_prime (Int64.of_int x)
*)
let is_prime = Misc.MillerRabin.is_prime_int

(* Given the list of numbers, return a list of numbers that are prime
   when a single digit is appended to the right. *)
let add_primes numbers =
  let result = ref [] in
  List.iter (fun number ->
    List.iter (fun extra ->
      let n = number * 10 + extra in
      if is_prime n then
	result := n :: !result
    ) [1; 3; 7; 9]
  ) numbers;
  !result

(* Generate a list of all right-truncatable primes. *)
let right_truncatable_primes () =
  let rec loop set result =
    (* printf "%s\n%!" (Show.show<int list> set); *)
    if set = [] then result else begin
      loop (add_primes set) (List.append set result)
    end in
  loop [2; 3; 5; 7] []

(* Is this number left truncatable? *)
let is_left_truncatable number =
  let rec loop number =
    if number = 0 then true
    else if number > 1 && is_prime number then
      loop (Misc.reverse_number (Misc.reverse_number number / 10))
    else false in
  loop number

let euler37 () =
  List.sum (List.filter (fun x -> x > 9 && is_left_truncatable x)
	      (right_truncatable_primes ()))

let run () = printf "%d\n" (euler37 ())
